### ISSUE TITLE
Better error messages

### ISSUE THREAD

**Issue by ofirgeller** [Microsoft/bond#769]
If possible when throwing "InvalidDataException" could you include info such as the property name or containing type?

This type of exception messages "'Invalid type 47, expected BT_LIST'" does not help much.

**Comment by chwarr** [comment]
Yes, type and field information could be forward on to the [method that actually generates](https://github.com/Microsoft/bond/blob/939c298ff6d619dbc0bc9f5ca25c2088de86d83b/cs/src/core/expressions/ThrowExpression.cs#L34-L37) the code that throws in many cases. This would be similar to what's done for `ThrowRequiredFieldMissingException`.

**Comment by ofirgeller** [comment]
@chwarr that would really help. for some reason fast binary works but simple binary throws an error on one of 200 possible properties.

**Comment by chwarr** [comment]
Makes sense.

My guess here is that the runtime schema being used to deserialize does not *exactly* match the schema that was used to serialize. When [using Simple Binary](https://microsoft.github.io/bond/manual/bond_cs.html#simple-binary), the (runtime) schema used when deserializing must be the schema that was used at serialization time. The Fast Binary and Compact Binary protocols have extra tags in the serialized payload that make it possible to use compatible--but not identical--schemas. Since Simple Binary lacks these tags, the two schemas must be identical.

**Comment by ofirgeller** [comment]
I am actually doing a simple test of round tripping the data to binary and back, and I am using the c# attributes to define the schema. So this must be a problem that happens even when the  schema is the same when serializing and deserializing. Might be that I added an attribute that is in conflict with the property it's on.
